### PR TITLE
revert mutex, since it introduces severe lock when gemm api for gemv i…

### PR DESCRIPTION
**Pull request by linlifan**
…s call with multithread

# Description
the unnecessary mutex introduce severe lock and performance issue, when calling gemm api for gemv with mutithread context.
suggest to get rid of it.

### CODE REVIEW

**Comment by dzarukin** [comment]
Hi @linlifan, what makes you think this mutex is unnecessary? What will happen if two threads reach this section simultaneously?

**Comment by linlifan** [comment]
@dzarukin  we have a case that call dnnl_threadpool_interop_sgemm to solve parallel gemv problem.  we observed that the mutex introduce many lock.
image

and from perf sche_switch , we can see that 
image

**Comment by dzarukin** [comment]
@linlifan, second look showed that `std::call_once` suites the intended service better - it still prevents multiple threads to override the final value with race condition happening, but also doesn't make the final value overridden at all unlike mutex which serializes the access for multiple threads. `call_once` should improve performance as well since threads will skip to the end of section once first thread copies the value. Could you make this replacement, please?

**Comment by linlifan** [comment]
> @linlifan, second look showed that `std::call_once` suites the intended service better - it still prevents multiple threads to override the final value with race condition happening, but also doesn't make the final value overridden at all unlike mutex which serializes the access for multiple threads. `call_once` should improve performance as well since threads will skip to the end of section once first thread copies the value. Could you make this replacement, please?

@dzarukin sure and addressed.

**Comment by dzarukin** [comment]
Thank you! Could you, please, squash all commits and rebase it on the top of the master branch?

**Comment by linlifan** [comment]
> Thank you! Could you, please, squash all commits and rebase it on the top of the master branch?

addressed. @dzarukin

File: src/cpu/x64/gemm/gemm_info.cpp [hunk 1]
```diff
@@ -16,7 +16,6 @@
 
 #include <cstdint>
 #include <memory>
-#include <mutex>
```
Comment (dzarukin) [inline src/cpu/x64/gemm/gemm_info.cpp:19]: This is still needed for `call_once`.

**Review by dzarukin — APPROVED** [review]
Please reword the commit:
`cpu: x64: gemm: replace mutex with call_once`
And additional message to explain that it allows to save some cycles would be appreciated.

File: src/cpu/x64/gemm/gemm_info.cpp (hunk 1 above)
Comment (linlifan) [reply src/cpu/x64/gemm/gemm_info.cpp]: Indeed.  addressed.

File: src/cpu/x64/gemm/gemm_info.cpp [hunk 2]
```diff
@@ -878,12 +876,8 @@ void gemm_info_t<a_t, b_t, c_t>::jit_init(void) {
     constexpr bool is_bf16 = data_traits<a_t>::data_type == data_type::bf16;
     bool doAlpha1 = this->alpha != 1.0f && is_bf16 ? no_alpha1 : do_alpha1;
 
-    {
-        // NOTE: This lock may not be needed at all as writes to copy_a_kern
-        // (and others) are protected within std::call_once(). The lock is added
-        // only to fix warnings reported by clang TSAN about a data race in
-        // this code block.
-        std::lock_guard<std::mutex> g(kern_mutex);
+    std::once_flag post_init_flag;
```
Comment (cantonios) [inline src/cpu/x64/gemm/gemm_info.cpp:879]: This will probably need to be `static` to actually guarantee it's only called once.

Comment (linlifan) [reply src/cpu/x64/gemm/gemm_info.cpp]: nice catch. addressed

**Comment by vpirogov** [comment]
Closing as stale.

**Pull request closed without merging**